src/scheduler/Executor.cpp

```diff
@@ -99,6 +116,189 @@ void Executor::finish()
+        } else {
+            exists = true;
+        }
+    }
+
+    if (exists) {
+        SPDLOG_DEBUG(
+          "Main thread snapshot exists: {} for {}", snapshotKey, funcStr);
+
+        // Get main snapshot
+        snap = reg.getSnapshot(snapshotKey);
+        std::span<uint8_t> memView = getMemoryView();
+
+        // Get dirty regions since last batch of threads
+        tracker.stopTracking(memView);
+        tracker.stopThreadLocalTracking(memView);
+
+        std::vector<std::pair<uint32_t, uint32_t>> dirtyRegions =
+          tracker.getBothDirtyOffsets(memView);
+
+        // Apply changes to snapshot
+        snap->fillGapsWithOverwriteRegions();
+        std::vector<faabric::util::SnapshotDiff> updates =
+          snap->diffWithDirtyRegions(memView, dirtyRegions);
+
+        if (updates.empty()) {
+            SPDLOG_TRACE(
+              "No updates to main thread snapshot for {} from {} dirty regions",
+              faabric::util::funcToString(msg, false),
+              dirtyRegions.size());
+        } else {
+            SPDLOG_DEBUG("Updating main thread snapshot for {} with {} diffs",
+                         faabric::util::funcToString(msg, false),
+                         updates.size());
+            snap->queueDiffs(updates);
+            snap->writeQueuedDiffs();
+        }
+
+        snap->clearMergeRegions();
+    }
+
+    // Now we have to apply the merge regions for this parallel section
+    for (const auto& mr : mergeRegions) {
+        snap->addMergeRegion(
+          mr.offset, mr.length, mr.dataType, mr.operation, true);
+    }
+
+    // TODO - here the main thread will wait, so technically frees up a slot
+    // that could be used.
+    std::string cacheKey =
+      std::to_string(msg.appid()) + "_" + std::to_string(req->messages_size());
+    bool hasCachedDecision = false;
+    {
+        faabric::util::SharedLock lock(threadExecutionMutex);
+        hasCachedDecision =
+          cachedDecisionHosts.find(cacheKey) != cachedDecisionHosts.end();
+    }
+
+    if (!hasCachedDecision) {
+        faabric::util::FullLock lock(threadExecutionMutex);
+        if (cachedDecisionHosts.find(cacheKey) == cachedDecisionHosts.end()) {
+            // Set up a new group
+            int groupId = faabric::util::generateGid();
+            for (auto& m : *req->mutable_messages()) {
+                m.set_groupid(groupId);
+                m.set_groupsize(req->messages_size());
+            }
+
+            // Invoke the functions
+            faabric::util::SchedulingDecision decision = sch.callFunctions(req);
+
+            // Cache the decision for next time
+            SPDLOG_DEBUG(
+              "No cached decision for {} x {}/{}, caching group {}, hosts: {}",
+              req->messages().size(),
+              msg.user(),
+              msg.function(),
+              groupId,
+              faabric::util::vectorToString<std::string>(decision.hosts));
+
+            cachedGroupIds[cacheKey] = groupId;
```

Comment:
I would beneffit from havng the decision cached in the scheduler, not in the executor. I am not asking you to change this, but writing this as a reminder to add a note in my PR that the logic is duplicated.